gradio/examples.py

```diff
@@ -0,0 +1,191 @@
+"""
+Defines helper methods useful for loading and caching Interface examples.
+"""
+from __future__ import annotations
+
+import csv
+import os
+import shutil
+from typing import TYPE_CHECKING, Any, Callable, List, Optional, Tuple
+
+from gradio.components import Dataset
+from gradio.flagging import CSVLogger
+
+if TYPE_CHECKING:  # Only import for type checking (to avoid circular imports).
+    from gradio import Interface
+    from gradio.components import Component
+
+CACHED_FOLDER = "gradio_cached_examples"
+
+
+class Examples:
+    def __init__(
+        self,
+        examples: List[Any] | List[List[Any]] | str,
+        inputs: List[Component],
+        outputs: Optional[List[Component]] = None,
+        fn: Optional[Callable] = None,
+        cache_examples: bool = False,
+        examples_per_page: int = 10,
+    ):
+        """
+        This class is a wrapper over the Dataset component can be used to create Examples
+        for Blocks / Interfaces. Populates the Dataset component with examples and
+        assigns event listener so that clicking on an example populates the input/output
+        components. Optionally handles example caching for fast inference.
+
+        Parameters:
+        examples (List[Any] | List[List[Any]] | str): example inputs that can be clicked to populate specific components. Should be nested list, in which the outer list consists of samples and each inner list consists of an input corresponding to each input component. A string path to a directory of examples can also be provided.
+        inputs: (Component | List[Component]): the component or list of components corresponding to the examples
+        outputs: (Component | List[Component] | None): optionally, provide the component or list of components corresponding to the output of the examples. Required if `cache` is True.
+        fn: (Callable | None): optionally, provide the function to run to generate the outputs corresponding to the examples. Required if `cache` is True.
+        cache_examples (bool): if True, caches examples for fast runtime. If True, then `fn` and `outputs` need to be provided
+        examples_per_page (int): how many examples to show per page (this parameter currently has no effect)
+        """
+        if not isinstance(inputs, list):
+            inputs = [inputs]
+        if not isinstance(outputs, list):
+            outputs = [outputs]
+
+        if examples is None or (
```

Comment:
`examples` cannot be None according to the type hint and when I set it to None the front-end hangs to load up the app in the blocks_input demo:

![image](https://user-images.githubusercontent.com/41651716/177397266-e2f3aa51-c3a5-4c64-abba-1b2b47361982.png)

If `None` is an allowed input we should figure out why the front-end can't display it. But if it's not (and I don't think it should be allowed btw) we should raise an appropriate error to the user?

Reply:
This is the repro:

```python
import gradio as gr
import os


def combine(a, b):
    return a + " " + b


def mirror(x):
    return x


with gr.Blocks() as demo:
    txt = gr.Textbox(label="Input", lines=2)
    txt_2 = gr.Textbox(label="Input 2")
    txt_3 = gr.Textbox("", label="Output")
    btn = gr.Button("Submit")
    btn.click(combine, inputs=[txt, txt_2], outputs=[txt_3])

    with gr.Row():
        im = gr.Image()
        im_2 = gr.Image()

    btn = gr.Button("Mirror Image")
    btn.click(mirror, inputs=[im], outputs=[im_2])

    gr.Markdown("## Text Examples")
    gr.Examples([["hi", "Adam"], ["hello", "Eve"]], [txt, txt_2], txt_3, combine, cache_examples=True)
    gr.Markdown("## Image Examples")
    gr.Examples(
        examples=None,
        inputs=im,
        cache_examples=False)

if __name__ == "__main__":
    demo.launch()
```